### PR TITLE
Keeps the 1/2->1/3->2/3 window behavior, but resets the cycle anytime…

### PR DESCRIPTION
**Pull request by kjrokos**
… a window is moved outside of the cycle

### CODE REVIEW

**Comment by rca** [comment]
Binary here: https://github.com/rca/ShiftIt/releases/tag/1.6.5

**Comment by rca** [comment]
Awesome, thank you!

Now, question is, what is the final desired behavior.  This is already better than the existing release, but some folks would like this feature disabled altogether.

I am not familiar at all with the code and don't know how easy or hard it is to add a toggle-able preference.

**Comment by kjrokos** [comment]
No problem, glad I could help!  I'm not really familiar with the entire codebase either`*`; the change above I made was pretty limited in scope -- I'm willing to take a look, but may take some time to familiarize myself with what I need to know. That said, it shouldn't be fairly simple, code wise, to add a toggle.  To do some of the other things people were requesting (user-defined split values, ignoring the second move command and then changing the split on the third, etc) would require a more serious refactoring of the code.

`*` or with Objective-c in general - my professional experience is C++ and Java

**Comment by fikovnik** [comment]
Amazing!

Regarding the settings, I think there are two ways how to implement it:

1. a quick one in which anyone who wants to disable it will have to run a command like `defaults write org.shiftitapp....halvesOnly`
2. add a GUI for it in the preference window - basically 

The first option involves:
- making a prefernce key (cf. https://github.com/fikovnik/ShiftIt/blob/40d2880bd6ea7c2eb555d1df35a6b3aaaa0c0c41/ShiftIt/ShiftItApp.h#L42 and https://github.com/fikovnik/ShiftIt/blob/40d2880bd6ea7c2eb555d1df35a6b3aaaa0c0c41/ShiftIt/ShiftItAppDelegate.m#L50)
- adding it to the plist file (cf. https://github.com/fikovnik/ShiftIt/blob/40d2880bd6ea7c2eb555d1df35a6b3aaaa0c0c41/ShiftIt/ShiftIt-defaults.plist) - editable easily with Xcode
- getting the actual value in the action (cf. https://github.com/fikovnik/ShiftIt/blob/master/ShiftIt/DefaultShiftItActions.m#L206)

**Comment by rca** [comment]
@kjrokos please comment on whether you think you'll have time to implement @fikovnik's suggestion above.  Thank you.

**Comment by kjrokos** [comment]
Yes, I can get you a PR in the next few days

File: ShiftIt/ShiftIt-defaults.plist [hunk 1]
```diff
@@ -85,14 +85,16 @@
 	<key>zoomKeyCode</key>
 	<integer>6</integer>
 	<key>zoomModifiers</key>
-    <integer>1835008</integer>
-   	<key>nextscreenKeyCode</key>
-   	<integer>45</integer>
-   	<key>nextscreenModifiers</key>
+	<integer>1835008</integer>
```
Comment (rca) [inline ShiftIt/ShiftIt-defaults.plist]: Is this an intentional whitespace change?

**Comment by kjrokos** [comment]
With the latest commit, you can now disable the functionality by using
```
defaults write org.shiftitapp.ShiftIt multipleActionsCycleWindowSizes NO
```

**Comment by rca** [comment]
Did you, by chance, make cycling default to enabled or disabled?  Given the backlash, perhaps it should be opt-in?

Also, adding the command in your comment above to the `README` is ideal.

I'm going to put together a new build ...

File: ShiftIt/ShiftIt-defaults.plist (hunk 1 above)
Comment (kjrokos) [reply ShiftIt/ShiftIt-defaults.plist]: More or less.  I used Xcode's plist editor, and it automatically made the whitespace changes; but had I edited it manually I would have likely done the same thing to make the file more consistent.

**Comment by rca** [comment]
Ok, just tried this myself and to enable cycling I had to run this command:

```
defaults write org.shiftitapp.ShiftIt multipleActionsCycleWindowSizes YES
```

**Comment by rca** [comment]
Adding reference to #254

**Comment by kjrokos** [comment]
Yeah, I wasn't sure what the default should be. I think it depends he backlash was because of the bug where going back and forth between directions changed the size (that was certainly the impetus for me getting involved), rather than the functionality itself.

I think for new installs, the default behavior will be for it to be enabled, because they don't have the default plist yet.  For upgrades, the default will be for it to be disabled, because the key doesn't exist in our plist.

**Comment by kjrokos** [comment]
(I didn't test the new install, though)

**Comment by rca** [comment]
That sounds reasonable.

If you don't mind updating the README, I think this is good to merge into `master`.  @fikovnik thoughts?

**Comment by jackjennings** [comment]
Can this get merged in soon? I don't know enough The new behavior that this pull request corrects bothers me every time I use the application…

**Comment by jackjennings** [comment]
Thanks so much @rca

**Comment by nmccready** [comment]
> With the latest commit, you can now disable the functionality by using
> 
> defaults write org.shiftitapp.ShiftIt multipleActionsCycleWindowSizes NO

Was this made the default (NO)? I only ask as the resizing used to work without the need to run.
`defaults write org.shiftitapp.ShiftIt multipleActionsCycleWindowSizes YES`

This is why I ended up on this thread.

**Comment by MattRyder** [comment]
Hey! You [broke my workflow!](https://xkcd.com/1172/)

I found the 'reduction cycle' feature recently, and thought this was default/intended behaviour.
I was about to file a bug report, glad the `defaults` functionality exists. All working on v1.6.6. 👍

**Comment by rca** [comment]
Per discussions I believe the intent was to maintain the old behavior and if someone wanted the additional 1/3 window in the flow the `defaults` command was required.

If this is not the case, please reply back and we should discuss maintaining old behavior as the default.

**Comment by concretevitamin** [comment]
As someone with auto update on, upgrading to 1.6.6 broke this behavior and I had to google, found this thread, and run `defaults write org.shiftitapp.ShiftIt multipleActionsCycleWindowSizes YES`.  I believe important -- and useful -- behaviors like this one should be kept backwards-compatible.  At least, make them visible in the GUI settings.

**Pull request merged**